packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts

```diff
@@ -235,21 +239,61 @@ export class OdspDocumentStorageService extends OdspDocumentStorageServiceBase {
 				{ eventName: "ObtainSnapshot", fetchSource },
 				async (event: PerformanceEvent) => {
 					const props: GetVersionsTelemetryProps = {};
-					let retrievedSnapshot: ISnapshotContents | undefined;
-
-					let method: string;
-					if (fetchSource === FetchSource.noCache) {
-						retrievedSnapshot = await this.fetchSnapshot(
+					let retrievedSnapshot:
+						| ISnapshotContents
+						| ISnapshotContentsWithEpoch
+						| undefined;
+
+					let method: string | undefined;
+					const snapshotCacheKey = createCacheSnapshotKey(this.odspResolvedUrl);
+					let prefetchedSnapshotP:
+						| Promise<ISnapshotContentsWithEpoch | undefined>
+						| Promise<ISnapshotContents>;
+					let inPrefetchNonPersistentCache = false;
+					// Only use the prefetched snapshot from non persistent cache in the load flow. Don't look into
+					// cache, if the host specifically tells us so.
+					if (!this.hostPolicy.avoidPrefetchSnapshotCache) {
+						const prefetchCacheKey = snapshotPrefetchCacheKeyFromEntry(snapshotCacheKey);
+						inPrefetchNonPersistentCache = this.cache.snapshotPrefetchResultCache?.has(prefetchCacheKey);
+						prefetchedSnapshotP = this.cache.snapshotPrefetchResultCache
+							?.get(prefetchCacheKey)
+							?.then(async (response) => {
+								// Validate the epoch from the prefetched snapshot result.
+								await this.epochTracker.validateEpoch(
+									response.fluidEpoch,
+									"treesLatest",
+								);
+								return response;
+							})
+							.catch(async (err) => {
+								this.logger.sendTelemetryEvent(
+									{
+										eventName: "PrefetchSnapshotError",
+										concurrentSnapshotFetch:
+											this.hostPolicy.concurrentSnapshotFetch,
+									},
+									err,
+								);
+								throw err;
+							}) ?? this.fetchSnapshot(
```

Comment:
I believe this flow breaks this.hostPolicy.concurrentSnapshotFetch === false promise.
The promise is - we will consult cache before making network call.
It feels like changes to logic should be not in this fuction, but rather be part of OdspDocumentStorageService.fetchSnapshot() flow.